### ISSUE TITLE
Linking error when registering masks or kernels during external module installation on macOS

### ISSUE THREAD

**Issue by TomBugnon** [nest/nest-simulator#841]
Dear NEST users, 

I am trying without success to install in nest-2.12.0 the example external module 'MyModule' extended with the 'EllipticMask' class (as described in the Topology user manual) and experience a linking error related to the registration of the mask with TopologyModule when compiling MyModule on Mac OS. I would greatly appreciate if anyone could help!

You can find the module and installation scripts necessary to replicate the issue, as well as the command line outputs of the nest and module installation on my machine at the following address: https://drive.google.com/drive/folders/0B0gyPoMB36PFLVNYNzBSU0MtNjA?usp=sharing

I am working on macOS Sierra and installed nest manually using the following commands:

```
cmake ${WORKING_DIR}/nest-2.12.0-src \
    -DCMAKE_INSTALL_PREFIX:PATH=${WORKING_DIR}/nest-2.12.0 \
    -DCMAKE_C_COMPILER=/usr/local/bin/gcc-7 \
    -DCMAKE_CXX_COMPILER=/usr/local/bin/g++-7 \
    -Dwith-python=3 \
    -Dwith-optimize=ON \
make -j8
make install
source ${WORKING_DIR}nest-2.12.0/bin/nest_vars.sh
```

Nest works nicely at that point, but I run into an error when I try installing my module using the commands below (as described in the Developer manual):
```
# From the directory containing MyModule
mkdir bld
cd bld
cmake ../MyModule \
    -Dwith-nest=$NEST_CONFIG_PATH  # path to nest-config.sh from nest install
make
make install
```

The configuration step runs fine but here is the error returned during the 'make' command:
```
[ 50%] Linking CXX shared library libmymodule.dylib
Undefined symbols for architecture x86_64:
  "nest::TopologyModule::mask_factory_()", referenced from:
      mynest::MyModule::init(SLIInterpreter*)     in mymodule.cpp.o
  "nest::Mask<2>::minus_mask(nest::AbstractMask const&) const", referenced from:
      vtable for mynest::EllipticMask in mymodule.cpp.o
  "nest::Mask<2>::union_mask(nest::AbstractMask const&) const", referenced from:
      vtable for mynest::EllipticMask in mymodule.cpp.o
  "nest::Mask<2>::intersect_mask(nest::AbstractMask const&) const", referenced from:
      vtable for mynest::EllipticMask in mymodule.cpp.o
  "nest::Mask<2>::inside(std::vector<double, std::allocator<double> > const&) const", referenced from:
      vtable for mynest::EllipticMask in mymodule.cpp.o
  "nest::Mask<2>::outside(nest::Box<2> const&) const", referenced from:
      vtable for mynest::EllipticMask in mymodule.cpp.o
ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [libmymodule.dylib] Error 1
make[1]: *** [CMakeFiles/mymodule_lib.dir/all] Error 2
make: *** [all] Error 2
```
You can find the full console output on the google drive.

The error is related to the registration of the new mask with in the module init and disappears if I comment out the corresponding line in mymodule.cpp ( 
  `` nest::TopologyModule::register_mask<EllipticMask>("elliptic"); ``). In which case I can install the module using nest.Install('mymodule') and use the neuron and synapse models defined in the module.

A few more elements to the puzzle:
- I encounter the same type of error ('undefined symbols for architecture x86_64' for a bunch of references) when trying to register a kernel with the nest::TopologyModule::register_parameter() function (if I try to add the affine2d kernel following the Topology User manual)
- I can install, import and use the example MyModule directory provided without this error (so the registration of neuron and synapse models works fine)
- I tried installing the module using the ``-Dexternal-modules`` cmake flag during nest installation without success.
- I tried the same installation process on a different mac machine without success.

Thank you so much for your help !

Tom Bugnon

**Comment by TomBugnon** [comment]
A few additions to the original issue:

- The ``nest::Mask`` symbols can be linked  when adding:

```
// Includes from topology:
#include "mask_impl.h"
```

to mymodule.cpp but the ``nest::TopologyModule::mask_factory_()`` error remains.

- I also included in the google drive the module ``RotatedRectangularMaskModule`` (which is the one I actually need). This module defines a topology mask that is rectangle (defined by its lower_left and upper_right coordinates) which is then rotated by a user-defined angle. The registration step for the rotated rectangular mask defined there raises the same type of error:
```
 Install:
Extracting help information; this may take a litte while.
[  0%] Built target generate_help
[ 25%] Linking CXX shared module rotatedrectangularmaskmodule.so
Undefined symbols for architecture x86_64:
  "nest::TopologyModule::mask_factory_()", referenced from:
      spiking_visnet::RotatedRectangularMaskModule::init(SLIInterpreter*)       in rotatedrectangularmaskmodule.cpp.o
  "nest::names::lower_left", referenced from:
      nest::AbstractMask* nest::GenericFactory<nest::AbstractMask>::new_from_dict_<spiking_visnet::RotatedRectangularMask>(lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype> const&) in rotatedrectangularmaskmodule.cpp.o
  "nest::names::upper_right", referenced from:
      nest::AbstractMask* nest::GenericFactory<nest::AbstractMask>::new_from_dict_<spiking_visnet::RotatedRectangularMask>(lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype> const&) in rotatedrectangularmaskmodule.cpp.o
ld: symbol(s) not found for architecture x86_64
```

The ``nest::names`` unresolved symbols are still not found by the linker when adding ``#include nest_names.h`` to rotatedrectangularmask.h or rotatedrectangularmask.cpp

Of note, the exact same RotatedRectangularMaskModule could be compiled, installed and used on a debian machine so there is a deeper issue somewhere.

- Finally I included in the drive a module trying to implement the `affine2d` kernel as described in the Topology User Manual. For this module I get the error:

```
 Install:
Extracting help information; this may take a litte while.
[  0%] Built target generate_help
[ 12%] Linking CXX shared library libmymodule.dylib
Undefined symbols for architecture x86_64:
  "nest::TopologyModule::parameter_factory_()", referenced from:
      mynest::MyModule::init(SLIInterpreter*)     in mymodule.cpp.o
  "nest::names::cutoff", referenced from:
      nest::TopologyParameter* nest::GenericFactory<nest::TopologyParameter>::new_from_dict_<mynest::Affine2DParameter>(lockPTRDatum<Dictionary, &SLIInterpreter::Dictionarytype> const&) in mymodule.cpp.o
```
However I didn't test this module on debian so there might be mistakes in the code.

### In summary 
The compiler can't find the mask_factory_(), the nest::names ( and possibly the parameter_factory) symbols on MacOS although at least some can be found on debian.

**Comment by stinebuu** [comment]
@TomBugnon sorry for the inactivity on this, is it still a problem?

**Comment by TomBugnon** [comment]
Hi, thanks for your answer. 
The issue still holds, yes. By the way I obtain a similar type of error when registering a neuron model.

I got around the problem by working from within a docker container where I can install modules without issue.

**Comment by TomBugnon** [comment]
My bad i deleted the files. You can find the install scripts and the modules here: 

https://drive.google.com/drive/folders/1Qhx4yJasnW5ejCRudCBKJfmEZIsU0_Mc?usp=sharing

Thanks again for your time, let me know if I can be of any help

**Comment by heplesser** [comment]
@TomBugnon Thanks for your files, I can reproduce the behaviour you see. One issue is simply resolved: in `rotatedrectangularmask.h`, you need to `#include "mask_impl.h"` in addition to `mask.h`. Problems with missing factories and names remain, though, I am working on it.
